### PR TITLE
Handle threads in thread groups 

### PR DESCRIPTION
**Pull request by skcin**
A thread handler is introduced to hold the threads in boost thread groups. This allows for the threads to be properly joined on shutdown and make the Qt wallet rebootable. 
Also the transaction database is closed on shutdown to prevent lock issues on restart.

It still takes too long to shutdown because of the ircseed thread. And not all threads in the wallet are moved to the thread handler.

### CODE REVIEW

**Review by tomasbrod — APPROVED** [review]
Well it compiles, loads, and when requested, shouts down. I am not familiar with this code so I don't know what to test more.

File: src/init.cpp [hunk 1]
```diff
@@ -232,6 +194,8 @@ bool AppInit(int argc, char* argv[])
 
     bool fRet = false;
 
+    boost::shared_ptr<ThreadHandler> threads(new ThreadHandler);
```
Comment (denravonska) [inline src/init.cpp:196]: We're moving away from boost so you can change this to `std::shared_ptr` from `<memory>`.

File: src/qt/bitcoingui.cpp [hunk 2]
```diff
@@ -728,17 +728,21 @@ void BitcoinGUI::createActions()
...
-	upgradeAction->setStatusTip(tr("Upgrade Client"));
-	upgradeAction->setMenuRole(QAction::TextHeuristicRole);
+    upgradeAction = new QAction(QIcon(":/icons/bitcoin"), tr("&Upgrade Client"), this);
+    upgradeAction->setStatusTip(tr("Upgrade Client"));
+    upgradeAction->setMenuRole(QAction::TextHeuristicRole);
 
-   aboutAction = new QAction(QIcon(":/icons/bitcoin"), tr("&About Gridcoin"), this);
-   aboutAction->setToolTip(tr("Show information about Gridcoin"));
-   aboutAction->setMenuRole(QAction::AboutRole);
+    rebootAction = new QAction(QIcon(":/icons/bitcoin"), tr("&Reboot Client"), this);
```
Comment (denravonska) [inline src/qt/bitcoingui.cpp:634]: Is the reboot option coming back?

Comment (denravonska) [reply src/qt/bitcoingui.cpp:634]: Removed it.

**Pull request merged**